File: code/packages/ruby/http_core/test/test_http_core.rb
# frozen_string_literal: true

require "minitest/autorun"
require "coding_adventures_http_core"

HC = CodingAdventures::HttpCore

class HttpCoreTest < Minitest::Test
  def test_version_round_trip
    version = HC::HttpVersion.parse("HTTP/1.1")
    assert_equal 1, version.major
    assert_equal 1, version.minor
    assert_equal "HTTP/1.1", version.to_s
  end

  def test_case_insensitive_header_lookup
    headers = [HC::Header.new("Content-Type", "text/plain")]
    assert_equal "text/plain", HC.find_header(headers, "content-type")
  end

  def test_content_helpers
    headers = [
      HC::Header.new("Content-Length", "42"),
      HC::Header.new("Content-Type", "text/html; charset=utf-8")
    ]
    assert_equal 42, HC.parse_content_length(headers)
    assert_equal ["text/html", "utf-8"], HC.parse_content_type(headers)
  end

  def test_heads_delegate_to_helpers
    request = HC::RequestHead.new("POST", "/submit", HC::HttpVersion.new(1, 1), [HC::Header.new("Content-Length", "5")])
    response = HC::ResponseHead.new(HC::HttpVersion.new(1, 0), 200, "OK", [HC::Header.new("Content-Type", "application/json")])

    assert_equal 5, request.content_length
    assert_equal ["application/json", nil], response.content_type
  end

  def test_body_kind_constructors
    assert_equal HC::BodyKind.new("none", nil), HC::BodyKind.none
    assert_equal HC::BodyKind.new("content-length", 7), HC::BodyKind.content_length(7)
    assert_equal HC::BodyKind.new("until-eof", nil), HC::BodyKind.until_eof
    assert_equal HC::BodyKind.new("chunked", nil), HC::BodyKind.chunked
  end
end

File: code/packages/python/http-core/tests/test_http_core.py
from http_core import (
    BodyKind,
    Header,
    HttpVersion,
    RequestHead,
    ResponseHead,
    find_header,
    parse_content_length,
    parse_content_type,
)


def test_http_version_round_trip() -> None:
    version = HttpVersion.parse("HTTP/1.1")
    assert version.major == 1
    assert version.minor == 1
    assert str(version) == "HTTP/1.1"


def test_find_header_is_case_insensitive() -> None:
    headers = [Header("Content-Type", "text/plain"), Header("X-Test", "ok")]
    assert find_header(headers, "content-type") == "text/plain"


def test_content_length_and_content_type_helpers() -> None:
    headers = [
        Header("Content-Length", "42"),
        Header("Content-Type", "text/html; charset=utf-8"),
    ]
    assert parse_content_length(headers) == 42
    assert parse_content_type(headers) == ("text/html", "utf-8")


def test_invalid_content_length_returns_none() -> None:
    assert parse_content_length([Header("Content-Length", "forty-two")]) is None


def test_heads_delegate_to_helpers() -> None:
    request = RequestHead(
        method="POST",
        target="/submit",
        version=HttpVersion(1, 1),
        headers=[Header("Content-Length", "5")],
    )
    response = ResponseHead(
        version=HttpVersion(1, 0),
        status=200,
        reason="OK",
        headers=[Header("Content-Type", "application/json")],
    )

    assert request.content_length() == 5
    assert response.content_type() == ("application/json", None)


def test_body_kind_constructors() -> None:
    assert BodyKind.none() == BodyKind("none")
    assert BodyKind.content_length(7) == BodyKind("content-length", 7)
    assert BodyKind.until_eof() == BodyKind("until-eof")
    assert BodyKind.chunked() == BodyKind("chunked")
